---
title: "Analyze data using ICARus"
author:
  - name: Zhaorong Li
    affiliation:
    - Boston University
output:
  html_document:
    self_contained: true
    toc: true
    toc_float:
      toc_collapsed: true
      toc_depth: 3
    fig_caption: yes
    code_folding: show
    theme: cerulean
editor_options: 
  chunk_output_type: console
package: "`r packageVersion('ICARus')`"
vignette: >
  %\VignetteIndexEntry{Introduction}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r chunk_setup, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)

bib <- list(
    fgsea = citation("fgsea")[1],
    ggplot2 = citation("ggplot2")[1],
    gridExtra = citation("gridExtra")[1],
    edgeR = citation("edgeR")[1],
    Rfast = citation("Rfast")[1],
    kneedle = citation('kneedle')
)
library(ICARus)
library(ggplot2)
library(gridExtra)
library(edgeR)
library(DT)
library(Rfast)

```



# Installation

Installing the ICARus package from Github using devtools.
```{r install from GitHub, eval=FALSE}
if (!require("devtools")) {install.packages('devtools')}
if (!require("ICARus")) {devtools::install_github('Zha0rong/ICARus',force = T)}
```

The following packages are recommended to be installed to explore the results from ICARus.

* fgsea `r knitcitations::citep(bib[["fgsea"]])`
* ggplot2 `r knitcitations::citep(bib[["ggplot2"]])`
* gridExtra `r knitcitations::citep(bib[["gridExtra"]])`
* edgeR `r knitcitations::citep(bib[["edgeR"]])`


```{r install Recommended Packages, eval=FALSE}
if (!require("fgsea")) {install.packages('fgsea')}
if (!require("ggplot2")) {install.packages('ggplot2')}
if (!require("gridExtra")) {install.packages('gridExtra')}
if (!require("edgeR")) {install.packages('edgeR')}
if (!require("DT")) {install.packages('DT')}
if (!require("Rfast")) {install.packages('Rfast')}
if (!require("ICARus")) {devtools::install_github('Zha0rong/ICARus',force = T)}


```

# Introduction

Independent Component Analysis (ICA) can be used to analyze transcriptome data in scenario where there is no definite categorical experiment settings. ICA can extract biologically meaningful gene expression signatures from the dataset, and these signatures can be associated with phenotype.

ICA is an unsupervised learning algorithm, which means that the output of the algorithm depends on the parameter being input. ICARus first identifies the near-optimal parameter for the ICA algorithm and runs the ICA algorithm using it, then ICARus runs ICA algorithm using different parameters and clusters the independent components extracted by different runs. The clustering results show how reproducible the signatures are.

In this vignette, we show how to analyze time-course RNA-Seq samples of Anti-CD3/CD28 activated T Cells and non-activated T Cells. The samples were obtained at 6 different time points. The oringal data is available [here](Anti-CD3/CD28 activation).

# Load Dataset

This RNA-Seq dataset contains 44 samples: 4 samples from 0 time point, 20 Anti-CD3/CD28 activate T Cells samples and 20 non-activated T Cells samples. 

```{r Load Dataset,error=FALSE,warning=FALSE,message=FALSE}

Count.Matrix=ICARus::Count.Matrix

metadata=ICARus::metadata

```

## Preview of metadata

The metadata of this RNA-Seq dataset is a data frame where the rows are samples and columns are phenotype information of the samples.

```{r Review metadata,error=FALSE,warning=FALSE,message=FALSE}

DT::datatable(metadata,filter = 'top',options = list(autoWidth = TRUE))

```

## Preview of Count Matrix

The Count Matrix of this RNA-Seq dataset is a matrix where the rows are gene symbols and columns are samples.

```{r Review Count.Matrix,error=FALSE,warning=FALSE,message=FALSE}
DT::datatable(head(Count.Matrix),filter = 'top',options = list(autoWidth = TRUE,scrollX = TRUE),width = '100%')
```

# Filtering and Normalizing Data

Genes that have 0 expression in more than half number of samples are removed.

The RNA-Seq dataset is Counts Per Million (CPM) normalized and log transformed using edgeR `r knitcitations::citep(bib[["edgeR"]])`.


```{r Normalizing,error=FALSE,warning=FALSE,message=FALSE}


Normalized=Count.Matrix
fun = function(x) {
  return(sum(x>=1)>=0.25*length(x))
}
rows.keep=apply(Normalized,1,FUN = fun)
Normalized=Normalized[rows.keep,]

Normalized=edgeR::cpm(Normalized)

Normalized=log(Normalized+1)

```

# Estimating the near-optimal paramter for ICA using Normalized Data.

Principal component analysis (PCA) is used to estimate the near-optimal parameter for ICA.
The elbow plot shows standard deviations of each principal components (PCs) in PCA, and the elbow point is the near-optimal parameter for ICA. In this dataset, the near-optimal parameter for ICA is 4. This means that ICA will extract at most 4 signatures from the data. 

```{r Estimation,error=FALSE,warning=FALSE,message=FALSE}

Estimation=ICARus::PCA.Estimation(Normalized)

Estimation$plot

```

# Running ICARus

`ICARus_complete()` function does the near-optimal parameter, signature robustness assessment and signature reproducibility automatically.

Based on the near-optimal parameter and the argument `numbers_of_parameter_for_reproducibility_test`, ICARus creates a set of parameters.

For every parameter in this set:

  fastICA is run 100 times (`iteration = 100`) using this parameter.
  
  The signatures from 100 runs are analyzed using hierarchical clustering (`clustering_algorithm = 'Hierarchical'`) based on the pearson correlations (`distance_measure = 'pearson'`) between signatures.
  FastICA algorithm extracts given number of signatures that are most independent of each other. The sizes of the clusters and the quality indices (icasso) from the clustering results are used to evaluate the robustness of the signatures (`quality.index.threshold`, `upperbound` and `lowerbound`).

After iterating through the parameters, the signatures that passed the robustness threshold from each parameter are analyzed using hierarchical clustering. Based on the size of the cluster one signature belongs to, the reproducibility of the signature among different parameters can be evaluate. The pipeline will output the signatures that are reproducible in more than half of the parameters tested.

```{r ICARus,error=FALSE,warning=FALSE,message=FALSE,results = "hide"}

Results=ICARus::ICARus_complete(Normalized,
                        iteration = 100,
                       numberofcores = 8,
                        numbers_of_parameter_for_reproducibility_test = 10,
                        distance_measure = 'pearson',
                        clustering_algorithm = 'Hierarchical',
                        tolerance = 1e-10,max.iteration = 10000,
                        upperbound = 100,lowerbound = 50,
                        quality.index.threshold = 0.75)

```

## Overview of results {.tabset}

The results of `ICARus_complete` are a list of objects:

  * `Quality_Index_of_Reproducible_Signatures`. This data frame reports how similar the signatures are to each other within the clusters.

  * `PCA_Elbow_Plot`. This plot is the elbow plot shown in previous section

  * `Raw.Results`. This list contains the raw results from fastICA runs.

  * `Reproducibility_Heatmap`. This heatmap visualizes the clustering of signatures from individual parameters.

  * `Reproducible_Signature_Matrix`. This matrix reports the final signatures extracted by ICARus pipeline. The rows are genes and columns are signatures.

  * `Reproducible_Affiliation_Matrix`. This matrix reports the final signatures extracted by ICARus pipeline. The rows are samples and columns are signatures.

### Quality_Index_of_Reproducible_Signatures

```{r Quality_Index_of_Reproducible_Signatures,error=FALSE,warning=FALSE,message=FALSE}

DT::datatable(Results$Quality_Index_of_Reproducible_Signatures,filter = 'top',options = list(autoWidth = TRUE,scrollX = TRUE),width = '100%')

```

### PCA_Elbow_Plot

```{r PCA_Elbow_Plot,error=FALSE,warning=FALSE,message=FALSE}

Results$PCA_Elbow_Plot

```

### Reproducibility_Heatmap

```{r Reproducibility_Heatmap,fig.width=15,fig.height=15,error=FALSE,warning=FALSE,message=FALSE}

Results$Reproducibility_Heatmap
              
```


### Reproducible_Signature_Matrix

```{r Reproducible_Signature_Matrix,error=FALSE,warning=FALSE,message=FALSE}

DT::datatable(head(Results$Reproducible_Signature_Matrix),filter = 'top',options = list(autoWidth = TRUE,scrollX = TRUE),width = '100%')
              
```


### Reproducible_Affiliation_Matrix

```{r Reproducible_Affiliation_Matrix,error=FALSE,warning=FALSE,message=FALSE}

DT::datatable(head(Results$Reproducible_Affiliation_Matrix),filter = 'top',options = list(autoWidth = TRUE,scrollX = TRUE),width = '100%')

```

# Interpreting Results from ICARus

## Associate signatures with phenotypes

The reproducible and stable gene signatures extracted by ICARus from the dataset have two parts: the signature matrix (`Reproducible_Signature_Matrix`) and the affiliation matrix (`Reproducible_Affiliation_Matrix`). The signature matrix has rows as genes and columns as signatures, and each signature can be viewed as ranked gene list and analyzed using GSEA. The affiliation matrix has rows as samples and columns as signatures, which allows users to associate signatures with phenotype.

In this example dataset, the two phenotypes of samples are time point and activation. One dot plot is generated for each signature where y axis is the signature score and x axis is the time point. The dot is colored by activation phenotype.


```{r Phenotype,error=FALSE,warning=FALSE,message=FALSE}


Phenotype=cbind(metadata,Results$Reproducible_Affiliation_Matrix)

DT::datatable(Phenotype,filter = 'top',options = list(autoWidth = TRUE,scrollX = TRUE),width = '100%')





```


```{r dot plot,fig.width=15,fig.height=15,error=FALSE,warning=FALSE,message=FALSE}

Signatures=list()

for (signature in colnames(Results$Reproducible_Affiliation_Matrix)) {
  plot=Phenotype[,c('Timepoint','Condition',signature)]
  colnames(plot)[3]='Signature.Score'
  plot=ggplot(plot,aes(x=Timepoint,y=Signature.Score,colour=Condition))+geom_point()+ggtitle(signature)
  Signatures[[signature]]=plot
  rm(plot,signature)
  
}
grid.arrange(grobs=Signatures)

```

## Interpreting gene signatures using GSEA

Each signature can be viewed as a ranked gene list, and hence can be analyzed using GSEA. 


```{r GSEA,fig.width=15,fig.height=15,error=FALSE,warning=FALSE,message=FALSE}
library(fgsea)
GSEA.Results=list()
GO=gmtPathways('https://data.broadinstitute.org/gsea-msigdb/msigdb/release/2023.2.Hs/c5.go.bp.v2023.2.Hs.symbols.gmt')


for (signature in colnames(Results$Reproducible_Signature_Matrix)) {
  name=signature
  signature=Results$Reproducible_Signature_Matrix[,signature]*(1)
  signature=signature[order(signature,decreasing = T)]
  
  signature.gsea=fgseaMultilevel(GO,signature,minSize = 15,maxSize = 500,eps = 0,nPermSimple = 10000)
  signature.gsea$leadingedgegene=''
  signature.gsea=signature.gsea[signature.gsea$padj<0.05,]
  
  for (i in 1:nrow(signature.gsea)) {
    signature.gsea$leadingedgegene[i]=paste(unlist(signature.gsea$leadingEdge[i]),collapse = ',')
  }
  
  signature.gsea$leadingEdge=NULL
  signature.gsea$pathway=gsub('^GOBP_',
                              '',
                              signature.gsea$pathway)
  signature.gsea$pathway=gsub('_',
                              ' ',
                              signature.gsea$pathway)
  
  signature.gsea$signature=name
  GSEA.Results[[name]]=signature.gsea
}

GSEA.Results=do.call(rbind,GSEA.Results)


DT::datatable(GSEA.Results,filter = 'top',options = list(autoWidth = TRUE,scrollX = TRUE),width = '100%')

```


# Session info {.unnumbered}

```{r sessionInfo, echo=FALSE}
sessionInfo()
```

